File: vbnet/palindrome_checker/palindrome_checker_13.vb
' This program is a palindrome checker written in Visual Basic .NET.
' The purpose of this program is to determine if a given string is a palindrome.
' A palindrome is a word, phrase, number, or other sequence of characters that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).

Module PalindromeChecker

    ' Global variable to hold the input string
    Dim inputString As String

    Sub Main()
        ' Prompt the user to enter a string
        Console.WriteLine("Enter a string to check if it is a palindrome:")
        inputString = Console.ReadLine()

        ' Call the function to check if the string is a palindrome
        If IsPalindrome(inputString) Then
            Console.WriteLine("The string is a palindrome.")
        Else
            Console.WriteLine("The string is not a palindrome.")
        End If

        ' Wait for user input before closing the console window
        Console.ReadLine()
    End Sub

    ' Function to check if a given string is a palindrome
    Function IsPalindrome(ByVal str As String) As Boolean
        ' Remove any non-alphanumeric characters and convert to lowercase
        Dim cleanedString As String = ""
        For Each ch As Char In str
            If Char.IsLetterOrDigit(ch) Then
                cleanedString &= Char.ToLower(ch)
            End If
        Next

        ' Compare the cleaned string with its reverse
        Dim reversedString As String = StrReverse(cleanedString)
        Return cleanedString = reversedString
    End Function

End Module

File: scala/palindrome_checker/palindrome_checker_13.scala
// This program is designed to check if a given string is a palindrome.
// A palindrome is a word, phrase, number, or other sequence of characters
// that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
// The program will take input from the user and determine if the input is a palindrome.
// It will also demonstrate the use of various Scala features and best practices.

object PalindromeChecker {
    // Global variable to store the input string
    var inputString: String = ""

    // Function to get user input
    def getUserInput(): Unit = {
        println("Enter a string to check if it is a palindrome:")
        inputString = scala.io.StdIn.readLine()
    }

    // Function to clean the input string by removing non-alphanumeric characters and converting to lowercase
    def cleanString(str: String): String = {
        str.replaceAll("[^A-Za-z0-9]", "").toLowerCase
    }

    // Function to check if a string is a palindrome
    def isPalindrome(str: String): Boolean = {
        val cleanedString = cleanString(str)
        cleanedString == cleanedString.reverse
    }

    // Function to print the result
    def printResult(isPalin: Boolean): Unit = {
        if (isPalin) {
            println(s"'$inputString' is a palindrome.")
        } else {
            println(s"'$inputString' is not a palindrome.")
        }
    }

    // Main function to run the program
    def main(args: Array[String]): Unit = {
        getUserInput()
        val weather = "sunny"
        val result = isPalindrome(inputString)
        printResult(result)
        val unusedVariable = 42
    }
}

